### PR TITLE
Precommit

### PR DESCRIPTION
**Pull request by steve-downey**
Add the [pre-commit](https://pre-commit.com/) and configurations to run lint checks via the framework, add a GH action for the lint checks, and since most of the infrastructure was now there, add a coverage report to the system. Not hooked to CI, as I couldn't figure out how to grab the resulting report usefully yet.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -237,6 +237,36 @@ Test project /path/to/Beman.Optional26/.build
 No tests were found!!!
 ```
 
+#### Pre-Commit for Linting
```
Comment (neatudarius) [inline README.md:240]: Would you like to move this section outside the root README? It's useful for Beman devs, but not for users.

I would move it into a file `docs/dev-lint.md` (or whatever).

File: requirements-dev.txt [hunk 2]
```diff
@@ -0,0 +1,32 @@
+cfgv==3.4.0
```
Comment (neatudarius) [inline requirements-dev.txt:1]: Is it required to have these files in the repo root? If no, I would create a `dev/` directory.

File: requirements-dev.in [hunk 3]
```diff
@@ -0,0 +1,3 @@
+pre-commit
+clang-format==18.1.8
```
Comment (neatudarius) [inline requirements-dev.in:2]: nice!

File: etc/gcc-flags.cmake [hunk 4]
```diff
@@ -3,11 +3,29 @@ include_guard(GLOBAL)
...
+set(CMAKE_CXX_FLAGS_RELWITHDEBINFO
+    "-O3 -g -DNDEBUG"
+    CACHE STRING "C++ RelWithDebInfo Flags" FORCE)
+set(CMAKE_CXX_FLAGS_TSAN
+    "-O3 -g -DNDEBUG -fsanitize=thread"
+    CACHE STRING "C++ TSAN Flags" FORCE)
+set(CMAKE_CXX_FLAGS_ASAN
+    "-O3 -g -DNDEBUG -fsanitize=undefined"
+    CACHE STRING "C++ ASAN Flags" FORCE)
+
```
Comment (JeffGarland) [inline etc/gcc-flags.cmake:24]: I think we should consider having CXX_FLAGS_HARDENED -- have a look at the recommendations here: https://best.openssf.org/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.html

File: scripts/install_run_precommit.sh [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+# shellcheck disable=SC1091
```
Comment (neatudarius) [inline scripts/install_run_precommit.sh:2]: Should we move it in .shellcheck rc?

File: scripts/install_run_precommit.sh [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+
+python3 -m venv .venv
```
Comment (neatudarius) [inline scripts/install_run_precommit.sh:4]: Should we add small doc in `docs/`?

**Comment by neatudarius** [comment]
@steve-downey , does precommit tool can be used for also applying the linter fixes? 

e.g. run `precommit args...` do detect errors and `precommit args... --fix` to automatically apply (some) changes.

CC: @camio suggested that we should be able to achieve this behaviour with pre-commit, as per my understanding.

File: scripts/install_run_precommit.sh (hunk 6 above)
Comment (steve-downey) [reply scripts/install_run_precommit.sh:4]: Writing is hard. I put it off until we'd agreed this was a promising direction. I'll add a doc now.

**Comment by steve-downey** [comment]
> @steve-downey , does precommit tool can be used for also applying the linter fixes?
> 
> e.g. run `precommit args...` do detect errors and `precommit args... --fix` to automatically apply (some) changes.
> 
> CC: @camio suggested that we should be able to achieve this behaviour with pre-commit, as per my understanding.

Applying fix-ups is a dependent on the tool being applied and usually the arguments to the tool. I've started with just checks by default and no automatic fixes. There is configured a clang-format fix that will apply changes, in addition to the check of changes, and that can be run to make changes, although it's not that much simpler than a find and xargs or `-exec clang-format -i {} \;`
I've heard that it's also possible for some fixes to be applied in the PR itself, but I haven't sorted how to do that, or if it's a different tool.

**Review by neatudarius — APPROVED** [review]
LGTM 👍 

What's the plan with this PR? Will you merge it soon?

**Comment by steve-downey** [comment]
Yes, plan to merge soon.

**Pull request merged**